src/github.com/matrix-org/dendrite/federationsender/consumers/roomserver_test.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumers
+
+import (
+	"testing"
+)
+
+func TestCombineNoOp(t *testing.T) {
+	inputAdd1 := []string{"a", "b", "c"}
+	inputDel1 := []string{"a", "b", "d"}
+	inputAdd2 := []string{"a", "d", "e"}
+	inputDel2 := []string{"a", "c", "e", "e"}
```

Comment:
This makes me unhappy. I'm reading the processing of this as:
```js
[a,b,c]  // add a b c
[c]  // del a b d
[a,c,d,e]  // add a d e
[d]  // del a c e e
```

This is further backed up by your comment for `combineDeltas`:

> Assumes that the order of operations is add(1), remove(1), add(2), remove(2).

So why are you expecting this to be `[]` and not `[d]`?